File: mock-facility-management/src/main/java/de/vkb/dojo/es/facilityManagement/services/commands/delegating/PickyCommandHandler.java
package de.vkb.dojo.es.facilityManagement.services.commands.delegating;

import de.vkb.dojo.es.facilityManagement.model.command.Command;
import de.vkb.dojo.es.facilityManagement.model.event.Event;
import de.vkb.dojo.es.facilityManagement.services.commands.CommandHandler;
import de.vkb.dojo.es.facilityManagement.services.commands.CommandHandlerResult;

public abstract class PickyCommandHandler<PC extends Command, C extends PC, PE extends Event, E extends PE> implements CommandHandler<PC, PE> {
    private final Class<C> clazz;

    public PickyCommandHandler(Class<C> clazz) {
        this.clazz = clazz;
    }

    public Boolean canHandle(PC command) {
        return clazz.isAssignableFrom(command.getClass());
    }

    public CommandHandlerResult<PC, PE> handle(PC command) {
        if (!canHandle(command)) {
            throw new IllegalArgumentException("cannot handle command of class " + command.getClass().getName());
        }
        //noinspection unchecked
        return (CommandHandlerResult<PC, PE>) doHandle((C) command);
    }

    protected abstract CommandHandlerResult<C, E> doHandle(C command);
}

File: mock-human-resources/src/main/kotlin/de/vkb/dojo/es/humanResources/services/commands/delegating/PickyCommandHandler.kt
package de.vkb.dojo.es.humanResources.services.commands.delegating

import de.vkb.dojo.es.humanResources.model.command.Command
import de.vkb.dojo.es.humanResources.model.event.Event
import de.vkb.dojo.es.humanResources.services.commands.CommandHandler
import de.vkb.dojo.es.humanResources.services.commands.CommandHandlerResult

abstract class PickyCommandHandler<PC : Command, C : PC, PE : Event, E : PE>(private val clazz: Class<C>) : CommandHandler<PC, PE> {
    fun canHandle(command: PC): Boolean {
        return clazz.isAssignableFrom(command.javaClass)
    }

    override fun handle(command: PC): CommandHandlerResult<PC, PE> {
        require(canHandle(command)) { "cannot handle command of class " + command.javaClass.name }
        @Suppress("UNCHECKED_CAST")
        return doHandle(command as C) as CommandHandlerResult<PC, PE>
    }

    protected abstract fun doHandle(command: C): CommandHandlerResult<C, E>
}
